slides/week4/week4.org

```diff
@@ -3,57 +3,336 @@
 #+EMAIL: kristaps@robojackets.org, rortiz9@gatech.edu, schintalapudi@gatech.edu
 
 * What are we doing today?
-- How a program's memory is structured at a high level
+- How a program's memory is structured and maintained
 - References and pointers
 - How to convert objects between different types
 - No =git= slides this week!
+- Pointer worksheets
+
+* What is the output?
+#+BEGIN_NOTES
+Ask them the output of this code. Then talk about how var and a are not the same place in memory.
+#+END_NOTES
+#+BEGIN_SRC c++
+#include <iostream>
+int increment(int a) {
+  a = a + 5;
+}
+int main() {
+  int var = 5;
+  increment(var);
+  std::cout << var << std::endl;
+}
+#+END_SRC
 
 * Memory structure
 #+BEGIN_NOTES
-This is all fundamentally the same RAM, just arranged and managed in a certain way. Static memory is used for static/global variables. A new stack frame is allocated when a function is called, and released when the function exits - this is where your parameters and local variables are stored. Mention memory leaks.
+Make sure to emphasize how important memory allocation is for any real program. This is all fundamentally the same RAM, just arranged and managed in a certain way. Static memory is used for static/global variables. A new block is allocated when a function is called, and released when the function exits - this is where your parameters and local variables are stored. Mention memory leaks.
 #+END_NOTES
 - At a high level, your program has three kinds of memory
- - *Static memory:* allocated at compile time
+ - *Static memory:* calculated at compile time
  - *The stack:* managed automatically by the language
  - *The heap:* your space to do whatever you want
-  - You can ask for memory from the OS and return it at runtime
+  - You can ask for memory from the OS and return it
+    - The heap is managed at runtime
   - You have to keep track of what you've allocated to avoid memory leaks
 - At an even higher level, all the memory in your computer is a huge, indexable, iterable space, divided into bytes
 
+** What is a byte
+#+BEGIN_NOTES
+They may never have seen binary before. 00101010 = 26, 00010001 = 17 do not talk about how binary works
+#+END_NOTES
+- 8 bits (0,1)
+  - 00101010
+  - 00010001
+- smallest indexable space on your computer
+
+** Object lifetimes stack
+#+BEGIN_NOTES
+Picture on next slide. talk about how in the below example the int is allocated on the stack. mention what stack overflow is
+#+END_NOTES
+- All variables have different lifetimes depending on where they are allocated in memory
+- Every method call has its own copy of the variables on the stack
+#+BEGIN_SRC c++
+#include <iostream>
+void increment(int a) {
```

Comment:
I may just be dumb but I don't see a difference between this code and the example that starts at line 73

Reply:
It is exactly the same but the picture did not fit so I pushed it to the next slide and copied the code. It happens a lot.